dot/sync/worker.go

```diff
@@ -1,95 +1,106 @@
-// Copyright 2021 ChainSafe Systems (ON)
+// Copyright 2023 ChainSafe Systems (ON)
 // SPDX-License-Identifier: LGPL-3.0-only
 
 package sync
 
 import (
-	"context"
-	"sync"
+	"errors"
+	"fmt"
 	"time"
 
-	"github.com/libp2p/go-libp2p/core/peer"
-
 	"github.com/ChainSafe/gossamer/dot/network"
-	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/libp2p/go-libp2p/core/peer"
 )
 
-// workerState helps track the current worker set and set the upcoming worker ID
-type workerState struct {
-	ctx    context.Context
-	cancel context.CancelFunc
+var ErrStopTimeout = errors.New("stop timeout")
 
-	sync.Mutex
-	nextWorker uint64
-	workers    map[uint64]*worker
+type worker struct {
+	status      byte
+	peerID      peer.ID
+	sharedGuard chan struct{}
+
+	punishment chan time.Duration
+	stopCh     chan struct{}
+	doneCh     chan struct{}
+
+	queue        chan *syncTask
+	requestMaker network.RequestMaker
 }
 
-func newWorkerState() *workerState {
-	ctx, cancel := context.WithCancel(context.Background())
-	return &workerState{
-		ctx:     ctx,
-		cancel:  cancel,
-		workers: make(map[uint64]*worker),
+func newWorker(pID peer.ID, sharedGuard chan struct{}, network network.RequestMaker) *worker {
+	return &worker{
+		peerID:       pID,
+		sharedGuard:  sharedGuard,
+		punishment:   make(chan time.Duration),
+		stopCh:       make(chan struct{}),
+		doneCh:       make(chan struct{}),
+		queue:        make(chan *syncTask, maxRequestsAllowed),
+		requestMaker: network,
+		status:       available,
 	}
 }
 
-func (s *workerState) add(w *worker) {
-	s.Lock()
-	defer s.Unlock()
-
-	w.id = s.nextWorker
-	w.ctx = s.ctx
-	s.nextWorker++
-	s.workers[w.id] = w
+func (w *worker) start() {
+	defer func() {
+		logger.Debugf("[STOPPED] worker %s", w.peerID)
+		close(w.doneCh)
+	}()
+
+	logger.Debugf("[STARTED] worker %s", w.peerID)
+	for {
+		select {
+		case <-w.stopCh:
+			return
+		case task := <-w.queue:
+			executeRequest(w.peerID, w.requestMaker, task, w.sharedGuard)
+		}
+	}
 }
 
-func (s *workerState) delete(id uint64) {
-	s.Lock()
-	defer s.Unlock()
-	delete(s.workers, id)
+func (w *worker) processTask(task *syncTask) (enqueued bool) {
+	select {
+	case w.queue <- task:
+		return true
+	default:
+		return false
+	}
```

Comment:
it may be simpler to just block on this.  What happens if we a worker is block indefinitely and we just continue looping in `syncWorkerPool.submitRequest` trying to assign this task to this queue. We could detect if a worker was indefinitely blocked using this select write default technique.